File: src-dotnet/RenesasForge.Protocol/Parsers/RForgeBinaryParser.cs
using RenesasForge.Core.Models;

namespace RenesasForge.Protocol.Parsers;

public sealed class RForgeBinaryParser : IProtocolParser
{
    private const byte Sof0 = 0xAA;
    private const byte Sof1 = 0x55;
    private const int HeaderSize = 8;
    private const int TailSize = 2;
    private const int MaxPayload = 1024;

    private readonly List<byte> _buffer = new();
    private readonly Queue<Frame> _frames = new();
    private ushort _nextSeq;

    public ParserType Type => ParserType.RForgeBinary;
    public int CrcErrorCount { get; private set; }
    public int OversizePayloadCount { get; private set; }

    public void Feed(ReadOnlySpan<byte> bytes)
    {
        _buffer.AddRange(bytes.ToArray());

        while (TryExtractNextFrame()) { }
    }

    public bool TryPopFrame(out Frame frame)
    {
        if (_frames.Count == 0)
        {
            frame = new Frame(CommandId.Unknown, 0, Array.Empty<byte>());
            return false;
        }

        frame = _frames.Dequeue();
        return true;
    }

    public byte[] BuildCommand(CommandId cmd, ReadOnlySpan<byte> payload)
    {
        var seq = _nextSeq++;
        var packet = new byte[HeaderSize + payload.Length + TailSize];
        packet[0] = Sof0;
        packet[1] = Sof1;
        packet[2] = 1;
        packet[3] = (byte)cmd;
        packet[4] = (byte)(seq & 0xFF);
        packet[5] = (byte)(seq >> 8);
        packet[6] = (byte)(payload.Length & 0xFF);
        packet[7] = (byte)(payload.Length >> 8);
        payload.CopyTo(packet.AsSpan(HeaderSize));

        var crc = Crc16Ccitt.Compute(packet.AsSpan(2, 1 + 1 + 2 + 2 + payload.Length));
        packet[HeaderSize + payload.Length] = (byte)(crc & 0xFF);
        packet[HeaderSize + payload.Length + 1] = (byte)(crc >> 8);
        return packet;
    }

    public void Reset()
    {
        _buffer.Clear();
        _frames.Clear();
        CrcErrorCount = 0;
        OversizePayloadCount = 0;
    }

    private bool TryExtractNextFrame()
    {
        if (_buffer.Count < HeaderSize + TailSize) return false;

        var sofIndex = FindSof();
        if (sofIndex < 0)
        {
            _buffer.Clear();
            return false;
        }

        if (sofIndex > 0) _buffer.RemoveRange(0, sofIndex);
        if (_buffer.Count < HeaderSize + TailSize) return false;

        var payloadLen = (ushort)(_buffer[6] | (_buffer[7] << 8));
        if (payloadLen > MaxPayload)
        {
            OversizePayloadCount++;
            _buffer.RemoveAt(0);
            return true;
        }

        var totalLen = HeaderSize + payloadLen + TailSize;
        if (_buffer.Count < totalLen) return false;

        var crcInput = new byte[1 + 1 + 2 + 2 + payloadLen];
        _buffer.CopyTo(2, crcInput, 0, crcInput.Length);
        var expected = (ushort)(_buffer[HeaderSize + payloadLen] | (_buffer[HeaderSize + payloadLen + 1] << 8));
        var actual = Crc16Ccitt.Compute(crcInput);
        if (expected != actual)
        {
            CrcErrorCount++;
            _buffer.RemoveAt(0);
            return true;
        }

        var cmd = (CommandId)_buffer[3];
        var seq = (ushort)(_buffer[4] | (_buffer[5] << 8));
        var payload = new byte[payloadLen];
        if (payloadLen > 0) _buffer.CopyTo(HeaderSize, payload, 0, payloadLen);
        _frames.Enqueue(new Frame(cmd, seq, payload));
        _buffer.RemoveRange(0, totalLen);
        return true;
    }

    private int FindSof()
    {
        for (var i = 0; i + 1 < _buffer.Count; i++)
        {
            if (_buffer[i] == Sof0 && _buffer[i + 1] == Sof1) return i;
        }

        return -1;
    }
}

File: src/protocol/RForgeBinaryParser.cpp
#include "protocol/RForgeBinaryParser.h"
#include "protocol/Crc16Ccitt.h"
namespace {
inline uint16_t readU16Le(const QByteArray& b, int i) {
    return static_cast<uint16_t>(static_cast<uint8_t>(b[i])) | static_cast<uint16_t>(static_cast<uint8_t>(b[i + 1]) << 8);
}
inline void appendU16Le(QByteArray& out, uint16_t v) {
    out.append(static_cast<char>(v & 0xFF));
    out.append(static_cast<char>((v >> 8) & 0xFF));
}
}
namespace rf {
void RForgeBinaryParser::feed(QByteArrayView bytes) {
    buffer_.append(bytes.data(), static_cast<int>(bytes.size()));
    while (buffer_.size() >= kHeaderSize + kTailSize) {
        int sof = -1;
        for (int i = 0; i + 1 < buffer_.size(); ++i) {
            if (static_cast<uint8_t>(buffer_[i]) == kSof0 && static_cast<uint8_t>(buffer_[i + 1]) == kSof1) { sof = i; break; }
        }
        if (sof < 0) { buffer_.clear(); return; }
        if (sof > 0) buffer_.remove(0, sof);
        if (buffer_.size() < kHeaderSize + kTailSize) return;

        const uint16_t payload_len = readU16Le(buffer_, 6);
        if (payload_len > kMaxPayload) { buffer_.remove(0, 1); ++crc_error_count_; continue; }

        const int total = kHeaderSize + payload_len + kTailSize;
        if (buffer_.size() < total) return;

        const QByteArray crc_input = buffer_.mid(2, 1 + 1 + 2 + 2 + payload_len);
        const uint16_t expect = readU16Le(buffer_, kHeaderSize + payload_len);
        const uint16_t actual = Crc16Ccitt::compute(QByteArrayView(crc_input));
        if (expect != actual) { buffer_.remove(0, 1); ++crc_error_count_; continue; }

        Frame frame;
        frame.cmd = static_cast<CommandId>(static_cast<uint8_t>(buffer_[3]));
        frame.seq = readU16Le(buffer_, 4);
        frame.payload = buffer_.mid(kHeaderSize, payload_len);
        queue_.enqueue(frame);
        buffer_.remove(0, total);
    }
}
bool RForgeBinaryParser::tryPopFrame(Frame& out) {
    if (queue_.isEmpty()) return false;
    out = queue_.dequeue();
    return true;
}
QByteArray RForgeBinaryParser::buildCommand(CommandId cmd, QByteArrayView payload) const {
    QByteArray out;
    out.reserve(kHeaderSize + static_cast<int>(payload.size()) + kTailSize);
    out.append(static_cast<char>(kSof0));
    out.append(static_cast<char>(kSof1));
    out.append(static_cast<char>(1));
    out.append(static_cast<char>(cmd));
    appendU16Le(out, 0);
    appendU16Le(out, static_cast<uint16_t>(payload.size()));
    out.append(payload.data(), static_cast<int>(payload.size()));
    const uint16_t crc = Crc16Ccitt::compute(QByteArrayView(out.mid(2)));
    appendU16Le(out, crc);
    return out;
}
void RForgeBinaryParser::reset() { buffer_.clear(); queue_.clear(); crc_error_count_ = 0; }
}
